### PR TITLE
defer Gloas full payload attestations until execution payload is available

### LINKED ISSUE #10720: Defer attestations that vote for FULL when the payload is not yet imported
## Problem

In Gloas, an attestation can signal that it is voting for the FULL variant of a beacon block. Today Teku can process that attestation once the beacon block root is known, even if the corresponding execution payload has not been imported yet and the FULL fork-choice node does not exist.

This creates a gap between "the beacon block is known" and "the FULL payload-backed node is available". Attestations in that gap should not be applied to fork-choice as if their target can already be resolved.

## Current Gap

Teku currently handles pending attestations for missing beacon block roots or timing-related cases, but not for:

```text
attestation votes for FULL
beacon block root is known
corresponding FULL payload/node is not yet available
```

This can cause a FULL vote to be applied before its actual fork-choice target exists.

## Expected Behavior

Attestations that depend on a FULL node should have an unresolved/pending flow when the corresponding payload has not yet been imported.

The attestation should not affect latest-message fork-choice weight until the FULL dependency is available.

As with the existing beacon block root dependency flow, recording this pending dependency could also trigger a by-root request for the missing payload/FULL data where the protocol supports it.

## Scope

This issue covers attestations received through gossip, local/REST submission, deferred same-slot processing, and attestations considered during block import.

The key requirement is source-independent:

```text
If an attestation votes for FULL, and the target FULL node is unavailable, do not apply it as a fork-choice vote yet.
```

## Acceptance Criteria

- FULL-payload attestations are not applied to fork-choice before the corresponding FULL node exists.
- Such attestations are either deferred until the FULL dependency is available or otherwise excluded without corrupting fork-choice accounting.
- The pending dependency can be used to trigger retrieval of the missing payload/FULL data, similarly to how missing beacon block roots are handled today.
- Replaying/reconsidering pending FULL attestations happens when the corresponding payload/FULL node becomes available.
- No FULL vote is silently applied to BASE or EMPTY as a fallback.
- Add regression coverage for a FULL attestation arriving before the payload, followed by payload import and later vote movement.

### PR DESCRIPTION
**Pull request by mehdi-aouadi**
## PR Description
- Fixes Gloas FULL attestations so they are deferred when the full execution payload is missing, instead of being applied to the base/empty fork choice node
- Adds `PendingAttestationPool` wrapping the existing pending attestation pool the same way `PendingBlockPool` wraps block pending state
- Reuses `PendingPool` for both missing block attestations and missing full payload votes
- `ForkChoice` now requests the missing full payload and replays deferred votes once the payload envelope is imported.

Rebased on top of #10804 

## Fixed Issue(s)

fixes #10720 

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

---

> [!NOTE]
> **Medium Risk**
> Changes fork-choice vote application and attestation deferral on a consensus-critical path; behavior is Gloas-specific but incorrect handling could affect head weight until payloads arrive.
> 
> **Overview**
> **Gloas full-payload attestations** are no longer applied to the empty/base fork-choice node when the full execution payload is missing. `ForkChoice` returns **`UNKNOWN_EXECUTION_PAYLOAD`**, skips applying those votes (including deferred and batched paths), and **`ForkChoiceModelGloas`** no longer falls back to the pending node when the full node is absent.
> 
> **`PendingAttestationPool`** (mirroring `PendingBlockPool`) holds attestations waiting on a **beacon block** or on a **full execution payload**, with slot/finalization handling moved out of `AttestationManager`. Parked full-payload votes are **reprocessed on block import** and on **`onExecutionPayloadImported`**.
> 
> **Beacon chain** wires the pool’s **required full payload** callback to **`RecentExecutionPayloadsFetcher`** so missing envelopes can be requested from peers. Tests cover the new defer/replay behavior and proto-array scoring when the full node is missing.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 1d61c1ab6f979466ea73500b4ede294c07f16e95. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/attestation/AttestationManager.java [hunk 1]
```diff
@@ -251,8 +249,12 @@ public SafeFuture<AttestationProcessingResult> onAttestation(
                   LOG.trace(
                       "Deferring attestation {} as required block is not yet present",
                       attestation::hashTreeRoot);
-                  pendingAttestations.add(attestation);
+                  pendingAttestationPool.addForMissingBlock(attestation);
                 }
+                case DEFERRED_FOR_EXECUTION_PAYLOAD ->
+                    LOG.trace(
+                        "Deferring attestation {} as required full payload is not yet present",
+                        attestation::hashTreeRoot);
```
[the remark this answers is not in the document]
Comment (mehdi-aouadi) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/attestation/AttestationManager.java]: fixed

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java [hunk 2]
```diff
@@ -1378,6 +1411,101 @@ private SafeFuture<Void> applyDeferredAttestations(final UInt64 slot) {
...
+            pendingVoteUpdatesBySlot
+                .computeIfAbsent(vote.slot(), VoteUpdates::new)
+                .addVote(vote.blockRoot(), vote.validatorIndex(), true));
+    final UInt64 currentSlot = spec.getCurrentSlot(recentChainData.getStore());
+    final List<DeferredVotes> readyVotes = new ArrayList<>();
+    pendingVoteUpdatesBySlot
+        .values()
+        .forEach(
+            votes -> {
+              if (votes.getSlot().isLessThan(currentSlot)) {
```
[the remark this answers is not in the document]
Comment (mehdi-aouadi) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java]: I don’t think this is a bug. `DeferredAttestations.prune(currentSlot)` uses `headMap(currentSlot, false)`, so it removes votes with slot `< currentSlot`.
If a payload arrives during slot N after `prune(N)` already ran, the vote is redeferred under slot N. It is not picked up immediately, but it is picked up on the next slot by `prune(N + 1)`, because slot N is `< N + 1`.
Applying it immediately at slot N would change the existing current slot attestation behaviour. I added a regression test covering this exact boundary case: payload arrives during the attestation’s current slot, the vote stays unapplied immediately then applies on the next slot tick

**Review by tbenr — COMMENTED** [review]
for the folks out there, we discussed this and agreed on a plan to steer it in a different direction

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java [hunk 3]
```diff
@@ -1378,6 +1393,58 @@ private SafeFuture<Void> applyDeferredAttestations(final UInt64 slot) {
...
+      final VoteUpdater voteUpdater,
+      final IndexedAttestationLight attestation) {
+    if (isFullPayloadVoteMissingTarget(
+        forkChoiceStrategy,
+        attestation.data().getBeaconBlockRoot(),
+        getFullPayloadVoteHint(attestation.data()))) {
+      return;
+    }
+    forkChoiceStrategy.onAttestation(voteUpdater, attestation);
+  }
```
[the remark this answers is not in the document]
Comment (mehdi-aouadi) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java:1430]: fixed

File: ethereum/statetransition/build.gradle [hunk 4]
```diff
@@ -16,6 +16,7 @@ dependencies {
 	implementation project(':infrastructure:async')
 	implementation project(':infrastructure:bls')
 	implementation project(':infrastructure:collections')
+	implementation project(':infrastructure:crypto')
```
Comment (tbenr) [inline ethereum/statetransition/build.gradle]: i believe this is not needed anymore

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java [hunk 5]
```diff
@@ -274,24 +278,41 @@ public SafeFuture<AttestationProcessingResult> onAttestation(
...
+                                  if (isFullPayloadVoteMissingTarget(
+                                      forkChoiceStrategy,
+                                      indexedAttestation.data().getBeaconBlockRoot(),
+                                      fullPayloadHint)) {
+                                    return AttestationProcessingResult.UNKNOWN_EXECUTION_PAYLOAD;
+                                  }
+                                  deferredAttestations.addAttestation(
+                                      indexedAttestation, fullPayloadHint);
+                                  return validationResult;
+                                });
```
Comment (tbenr) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java:298]: here we could early check the `fullPayloadHint` and only run `onForkChoiceThread` if voting on full (which requires a protoarray lookup via `isFullPayloadVoteMissingTarget`)

```
  --- a/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java
  +++ b/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java
  @@ onAttestation(...)
                             if (validationResult.getStatus() == Status.DEFER_FORK_CHOICE_PROCESSING) {
                               final IndexedAttestationLight indexedAttestation =
                                   getIndexedAttestation(attestation);
  +                            final boolean fullPayloadHint =
  +                                getFullPayloadVoteHint(attestation.getData());
  +                            if (!fullPayloadHint) {
  +                              // No full payload target to check (pre-Gloas attestation or
  +                              // empty-payload vote), so stay off the fork choice thread and
  +                              // batch the vote as before
  +                              deferredAttestations.addAttestation(indexedAttestation, false);
  +                              return SafeFuture.completedFuture(validationResult);
  +                            }
                               return onForkChoiceThread(
                                   () -> {
  -                                  final boolean fullPayloadHint =
  -                                      getFullPayloadVoteHint(attestation.getData());
  -                                  final ForkChoiceStrategy forkChoiceStrategy =
  -                                      getForkChoiceStrategy();
                                     if (isFullPayloadVoteMissingTarget(
  -                                      forkChoiceStrategy,
  +                                      getForkChoiceStrategy(),
                                         indexedAttestation.data().getBeaconBlockRoot(),
  -                                      fullPayloadHint)) {
  +                                      true)) {
                                       return AttestationProcessingResult.UNKNOWN_EXECUTION_PAYLOAD;
                                     }
  -                                  deferredAttestations.addAttestation(
  -                                      indexedAttestation, fullPayloadHint);
  +                                  deferredAttestations.addAttestation(indexedAttestation, true);
                                     return validationResult;
                                   });
                             }
                             return SafeFuture.completedFuture(validationResult)

```

**Review by tbenr — DISMISSED** [review]
LGTM with just 2 things.

(you should remove the `PendingFullPayloadVote` mention in the PR description)

File: ethereum/statetransition/build.gradle (hunk 4 above)
Comment (mehdi-aouadi) [reply ethereum/statetransition/build.gradle]: Forgot to remove it which makes me think that it would great to have something that detects such things. Should be the bug bot job imo

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java (hunk 5 above)
Comment (mehdi-aouadi) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java:298]: good catch, addressed it

**Review by tbenr — APPROVED** [review]
LGTM
